### PR TITLE
Persistent networks and checkpoint support in z2

### PR DESCRIPTION
**Pull request by rrw-zilliqa**
(feat) The ability to specify network shapes, save them across z2 runs, and start them in groups.

(feat) The ability to load checkpoints from z2

Relates to #1364

### CODE REVIEW

File: z2/src/setup.rs [hunk 1]
```diff
@@ -48,73 +65,164 @@ pub struct Setup {
...
+            } else {
+                Config::from_spec(val, base_port)?
+            }
+        } else if let Some(val2) = loaded_config {
+            println!("Starting previously saved config in {config_dir}");
+            val2
+        } else {
+            // Set up a default network
+            println!(">> No network specified or loaded; using default 4-node network for legacy reasons.");
+            //config = loaded_config.ok_or(anyhow!(
```
Comment (86667) [inline z2/src/setup.rs:183]: can be removed?

File: z2/src/node_spec.rs [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+    pub is_validator: bool,
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize, Default)]
+pub struct Composition {
+    pub nodes: HashMap<u64, NodeDesc>,
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize, Default)]
+pub struct NodeSpec {
```
Comment (86667) [inline z2/src/node_spec.rs:19]: I wonder if `NetworkSpec` or `NodesSpec` is more clear here

Comment (rrw-zilliqa) [reply z2/src/node_spec.rs:19]: I'll leave that to the language lawyers - when those who care enough have decided what they want, very happy to implement it :-)

File: z2/src/setup.rs (hunk 1 above)
Comment (86667) [reply z2/src/setup.rs]: nope, i was talking about the commented out code :+1:

**Comment by 86667** [comment]
Nice! Will be handy.

It works well on initialising a new network but has trouble when a config already exists. I get:

```
WARNING: You've specified a network configuration; we'll ignore this and take the config from the existing loaded configuration file
```

but it does later attempt to read a node that doesnt exist, so the nodespec config is infact being used as well as the current config. 

I will approve anyway as its an improvement nevertheless

**Comment by rrw-zilliqa** [comment]
> Nice! Will be handy.
> 
> It works well on initialising a new network but has trouble when a config already exists. I get:
> 
> ```
> WARNING: You've specified a network configuration; we'll ignore this and take the config from the existing loaded configuration file
> ```
> 
> but it does later attempt to read a node that doesnt exist, so the nodespec config is infact being used as well as the current config.
> 
> I will approve anyway as its an improvement nevertheless (once its out of draft)

Thanks! I'll rerequest when the clilppy checks have passed - in the meantime, can  you give me a set of commands to reproduce your issue and I'll fix it? It seems like something I ought to fix ...

**Review by 86667 — APPROVED** [review]

**Pull request merged**